### DISCUSSION TITLE
Using MARKER within a colon definition?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#133] [Q&A] [answered]
I'd like to define a word that sets a marker at the start of a group of tests and clean-up those tests at the end of the group.  Something like:

```
: test_group ... MARKER rm_test_group ... ;
: test_group_end ... rm_test_group ;
```
However `MARKER` wants to add a word to the word list, which makes it problematic while in the middle of a word definition.  Is there some trick or method I'm missing that could achieve this?  Or is it just not possible?

**Comment by paraplegic** [comment]
Perhaps simply using the HERE word to put the current address on the 
stack?

On Thu, 26 Jan 2023, Anthony Howe wrote:

> Date: Thu, 26 Jan 2023 12:05:08 -0800
> From: Anthony Howe ***@***.***>
> Reply-To: ForthHub/discussion
>     ***@***.***>
> To: ForthHub/discussion ***@***.***>
> Cc: Subscribed ***@***.***>
> Subject: [ForthHub/discussion] Using MARKER within a colon definition?
>     (Discussion #133)
> 
> 
> I'd like to define a word that sets a marker at the start of a group of tests and clean-up those tests at the end of the group. Something like:
> 
> : test_group ... MARKER rm_test_group ... ;
> : test_group_end ... rm_test_group ;
> 
> However MARKER wants to add a word to the word list, which makes it problematic while in the middle of a word definition. Is there some trick or method I'm missing that could achieve
> this? Or is it just not possible?
> 
> ?
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.[AAS2KEUZMYKW4XC4VEUWCWLWULKHJA5CNFSM6AAAAAAUH5BVMSWGG33NNVSW45C7OR4XAZNKIRUXGY3VONZWS33OVJRW63LNMVXHIX3JMTHAASJN2U.gif]
> Message ID: ***@***.***>
> 
> 
>

-- 
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
Robert S. Sciuk                                         ***@***.***
Principal Consultant            Phone: 289.312.1278   Cell: 905.706.1354
Control-Q Research                  97 Village Rd. Wellesley, ON N0B 2T0

**Reply by SirWumpus** [reply]
I don't see how that would help / work.

**Comment by ruv** [comment, marked as answer]
A simple solution is to use `evaluate`:
```
: test_group ... "marker rm_test_group" evaluate ... ;
: test_group_end ... "rm_test_group" evaluate  ;
```

Or [`execute-parsing`](https://github.com/forthy42/gforth/blob/master/compat/execute-parsing.fs) in the first case:
```
: test_group "rm_test_group" ['] marker execute-parsing ... ;
```

-----

The original code:
> ```
> : test_group ... MARKER rm_test_group ... ;
> : test_group_end ... rm_test_group ;
> ```

is not correct due to the following problems:

1. The word `marker` is an ordinary word. Therefore it doesn't parse anything when it's used inside a definition (i.e., when it's compiled), — just like `constant` and `:` (colon). So the Forth text interpreter will encounter `rm_test_group` and raise an error.
2. When `test_group_end` is being defined, `rm_test_group` has not been defined yet, so an error will be raised.

**Reply by SirWumpus** [reply]
Ah.  I knew I'd forgotten something, like `evaluate`.  That method works nicely and seems the cleanest expression.

`execute-parsing` isn't a standard word or implies a particular implementation, which is less portable.

**Comment by gerryjackson** [comment]
If I understand what you're asking for you could do:

defer do-marker

: test-group >in @ marker >in ! ' ['] do-marker defer! ;

: test-group-end do-marker  ;

\ Example

wordlist constant mk-wl    \ to see tests defined and removed
mk-wl >order definitions
test-group foo
: test1 ( ... ) ;
: test2 ( ... ) ;
: test3 ( ... ) ;
words
test-group-end
words

\ Works in Gforth

Gforth 0.7.9_20180905, Copyright (C) 1995-2017 Free Software Foundation, 
Inc.
Gforth comes with ABSOLUTELY NO WARRANTY; for details type `license'

defer do-marker
  : test-group >in @ marker >in ! ' ['] do-marker defer! ;
: test-group-end do-marker  ;

  \ Example
Gforth 0.7.9_20180905, Copyright (C) 1995-2017 Free Software Foundation, 
Inc.
Gforth comes with ABSOLUTELY NO WARRANTY; for details type `license'
Type `help' for basic help

wordlist constant mk-wl    \ not needed - to see tests defined and removed
mk-wl >order definitions
test-group foo
: test1 ( ... ) ;
: test2 ( ... ) ;
: test3 ( ... ) ;
words      \ Display definitions in mk-wl
test3 test2 test1 foo
test-group-end
words       \ Ditto - test group removed

\ Alternatively more simply

: test-group marker ;
: test-group-end  ' execute  ;

\ Usage

test-group foo
\ Test code as above
end-test-group foo

**Reply by ruv** [reply]
@gerryjackson, nice solution!

Though, as I can see, the idea of Anthony was to have a kind of anonymous lexical blocks of tests:
```
test_group
  ...
test_group_end
```

They also work as expected if they are nested.
NB: since `test_group_end` is a word that should be encountered by the Forth text interpreter, the contained tests shall not shadow this word or make it unfindable (due to changes in the search order).

**Comment by gerryjackson** [comment]
I realised soon after I'd posted, that in my first solution the word 
do-marker is unnecesssary. Just have

defer test-group-end

and replace do-marker in the definition of test-group with test-group-end.
